README
INTRODUCTION
  
  The sandbox libraries (libsandbox & pysandbox) provide API's in C/C++ and
  Python for executing and profiling simple (single process) programs in a 
  restricted environment, or sandbox. These API's can help developers to build
  automated profiling tools and watchdogs that capture and block the runtime 
  behaviours of binary executable programs according to configurable / 
  programmable policies.
  
  The sandbox libraries are distributed under the terms of the New BSD license, 
  please refer to the plain text file named COPYING in individual packages.
  
  Project Homepage: http://openjudge.net/~liuyu/Project/LibSandbox
  Author: LIU Yu, <pineapple.liu@gmail.com>

FEATURES
  
  1. capture system calls with arguments from *sandboxed* binary programs in 
     runtime, and block malicious actions through user-defined policy modules;
  2. specify quota limit of resources allocated to the *sandboxed* program, 
     including cpu and wallclock time, memory, and disk output;
  3. minimize privileges of the *sandboxed* program, and isolate its execution
     from critical parts of the operating system;

INSTALLATION
  
  Note, there is no need to install 32bit package(s) on x86_64 systems. The 
  64bit libsandbox can handle both 32bit and 64bit binary programs.
  
  binary package (CentOS/RedHat)
  
  $ yum install --nogpgcheck libsandbox-<version>.<platform>.rpm
  
  binary package (Ubuntu)
  
  $ dpkg -i libsandbox_<version>_<platform>.deb
  
  source package
  
  $ tar -xzf libsandbox-<version>.tar.gz
  $ cd libsandbox-<version>
  $ ./configure
  $ sudo make install

SPECIAL NOTICES
  
  1. libsandbox is designed for, and has been tested on, linux-2.6 (CentOS 6.2,
     Ubuntu 11.4) and linux-3.0 (Ubuntu 11.10, Ubuntu 12.10) systems on i686 
     and x86_64 architectures (x86_64 supported since v0.3.x). The behaviours 
     of libsandbox on other platforms are undefined, even if, though unlikely, 
     they pass configuration and (or) compilation;
  2. As of v0.3.5, libsandbox only supports single process ELF programs. If the
     user-provided policy module bypasses system calls that spawn new processes
     (i.e. returning S_ACTION_CONT upon seeing SYS_fork, SYS_vfork, SYS_clone)
     the subsequent behaviours of libsandbox are undefined;
  3. On x86_64 systems, both 32bit and 64bit programs MUST be *sandboxed* by  
     64bit libsandbox. In particular, NEVER run any program (not even native
     32bit programs) with 32bit libsandbox on 64bit systems. Otherwise, there 
     is suspected risk that a malicious 32bit program can bypass some system 
     call restrictions through mapping in 64bit system call table;
  4. As of v0.3.5, libsandbox internally uses ptrace() to intercept and probe 
     system calls invoked by the *sandboxed* program. If the *sandboxed* program
     itself invokes ptrace(), and that the user-provided policy module bypasses 
     relevant events (i.e. returning S_ACTION_CONT upon seeing SYS_ptrace), the 
     subsequent behaviours of libsandbox are undefined;
  5. libsandbox interprets some signals sent to the *sandboxed* program as out- 
     of-quota (wallclock, cpu, memory, disk) events (for v0.2.x the affiliated 
     signals are SIGXCPU, SIGXFZ, SIGPROF, SIGALRM, SIGVTALRM, and SIGUSR1; for
     v0.3.x the only affiliated signal is SIGXFSZ). If the *sandboxed* program 
     depends on these signals to perform internal functionalities, the user-
     provided policy module of libsandbox may cause the *sandboxed* program to 
     terminate upon seeing relevant events. Moreover, if user-provided policies
     bypass system call events that block such signals (i.e. returning S_ACTION
     _CONT without checking parameters of SYS_sigsetmask, SYS_pthread_sigmask, 
     etc.), libsandbox could fail to detect relevant out-of-quota events;
  6. By default, the disk output quota only applies to persistent storage. 
     Write operations to i) standard output buffer (i.e. stdout / stderr), ii) 
     null devices (i.e. /dev/null), or iii) pipelined to another process do NOT 
     count against the disk output quota. This default behaviour, however, 
     could be overridden by user-specified policy modules.
  7. Since v0.3.5, libsandbox dedicates a manager thread to handle termination 
     signals including SIGTERM, SIGINT, and SIGQUIT. When the manager thread
     receives any of these signals, it sends the same signal to the *sandboxed*
     program. But the supervisor process, or thread, running libsandbox may 
     continue to execute if i) the *sandboxed* program is still running after 
     receiving the signal, and ii) the user-specified policy module explicitly 
     requires libsandbox to continue;
  8. If the supervisor process, or thread, running libsandbox (i.e. which is 
     the parent process of the *sandboxed* program) terminates upon receiving 
     termination signals other than SIGTERM, SIGINT, and SIGQUIT, or if custom 
     signal handlers are registered such that they prevent the manager thread 
     of libsandbox from receiving the three termination signals, then there is
     suspected risk that the *sandboxed* program may go beyond the control of 
     libsandbox. In such cases, some of the restrictions placed by libsandbox,
     including quota limit and policy-based behaviour auditing, may become 
     invalid. However, OS-level security mechanisms, including chroot() jail 
     and setuid() privileges, are still in effect.
  9. libsandbox (v0.3.x) includes some optional features that can be enabled
     during configuration. Please note that --enable-tsc and --enable-rtsched 
     are highly experimental, and are not recommended for production systems;
 10. On some linux-2.6-x86_64 systems, to correctly inspect the system call mode
     of some 32bit programs, it could be mandatory to build libsandbox (0.3.x) 
     with --enable-chkvsc option (since 0.3.3-rc4). The binary packages for 
     x86_64 systems enabled this option by default;

CONTACT
  
  Project Homepage: http://openjudge.net/~liuyu/Project/LibSandbox
  Project Page at SourceForge.net: http://sourceforge.net/projects/libsandbox/
  Project Page at Google Code: http://code.google.com/p/libsandbox/
  Project Mail List at Google Groups: http://groups.google.com/group/open-judge
  Author: LIU Yu, <pineapple.liu@gmail.com>


pysandbox/CHANGES
[2013/04/12] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c revised Sandbox_dump() to suppress some aggressive 
    exceptions when failing to dump data due to invalid address

[2013/04/10] LIU Yu, <pineapple.liu@gmail.com>
  * in setup.{py,cfg} updated version to 0.3.5-3

[2013/03/24] LIU Yu, <pineapple.liu@gmail.com>
  * in setup.py added -lm to linker arguments

[2013/01/03] LIU Yu, <pineapple.liu@gmail.com>
  * in setup.{py,cfg} enabled static linkage against the core library

[2012/10/15] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c fixed clear order of objects
  * in setup.{py,cfg} updated version to 0.3.5-2

[2012/09/04] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c fixed early unlock in Sandbox_traverse()

[2012/08/29] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c optimized string dump in Sandbox_dump()
  * in sandbox/module.c removed redundant exception generation

[2012/08/20] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c applied new LOCK/UNLOCK macros for concurrency control
  * in sandbox/module.c changed status and result attributes of Sandbox from
    instance members to getter methods
  * in setup.{py,cfg} updated version to 0.3.5-1

[2012/07/19] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c added new attributes owner and group to Sandbox

[2012/07/12] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c updated signal blocking due to changes in libsandbox

[2012/07/10] LIU Yu, <pineapple.liu@gmail.com>
  * in README, setup.py sandbox/__init__.py updated introductive message
  * in sandbox/module.c block signals used by libsandbox during module init

[2012/07/09] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.h added io field in Sandbox C-structure
  * in sandbox/module.c Sandbox now holds (incref) input / output file objects
    to ensure their validity during the lifetime of the sandbox object
  * in sandbox/module.c optimized SandboxPolicy_entry method
  * in sandbox/module.c added new Get attribute quota to Sandbox
  * in sandbox/module.c fixed an off-by-one error in Sandbox_load_quota
  * in sandbox/module.h moved all doc strings to sandbox/module.c
  * in sandbox/module.c wrapped all doc strings with PyDoc_STRVAR()

[2012/07/08] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c optimized SandboxPolicy call method
  * in santbox/module.c added cyclic GC support for Sandbox and SandboxPolicy
  * in sandbox/module.{h,c} added a new generic base type anyType

[2012/07/07] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c fixed refcount error in case of Sandbox init failure
  * in sandbox/module.c added new Get attributes jail, task, pid and GetSet
    attribute policy to Sandbox
  * in sandbox/module.c changed accessibility of SandboxEvent members to RO
  * in sandbox/module.c unified naming of global variables
  * in sandbox/module.h changed error type of policy check failure from Assertion-
    Error to ValueError and updated respective error messages
  * in sandbox/module.h removed repetitive doc strings of Sandbox methods
  * in setup.py query pkg-config for compile / link arguments of the core library
  * in setup.{py,cfg} updated version to 0.3.4-3

[2012/07/04] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/__init__.py converted EOL to UNIX conventions
  * in sandbox/module.h revised some error messages
  * in sandbox/module.c moved PY3K detection macros to sandbox/module.h
  * in sandbox/module.c renamed SandboxQuota_FromPyObject() -> ..._FromObject()
  * in setup.{py,cfg} updated version to 0.3.4-2

[2012/06/08] LIU Yu, <pineapple.liu@gmail.com>
  * done with py3k compatibility

[2012/06/03] LIU Yu, <pineapple.liu@gmail.com>
  * refactored code and compile args for C++ compatibility
  * refactored code for py3k compatibility
  * removed support for str()/unicode() quota values

[2012/04/25] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c moved inline help messages to sandbox/module.h
  * wrapped sandbox/module.{h,c} with extern "C" for C++ compatibility
  * in sandbox/module.h include <Python.h> before standard headers to suppress
    warnings on _POSIX_C_SOURCE redefinition

[2012/04/22] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.h removed bool definition and in sandbox/module.c added 
    <sandbox-dev.h> inclusion
  * in sandbox/module.c revised affiliated code for py3k forward compatibility
  * in setup.{py,cfg} updated version to 0.3.4-1

[2012/04/03] LIU Yu, <pineapple.liu@gmail.com>
  * in setup.{py,cfg} updated version to 0.3.3-rc5

[2012/03/27] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c removed SandboxPolicy_default_policy() and applied
    sandbox_default_policy() from libsandbox

[2012/03/26] LIU Yu, <pineapple.liu@gmail.com>
  * in setup.cfg stated conflict with pysandbox-ext
  * in setup.{py,cfg} updated version to 0.3.3-rc4

[2011/12/15] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c fixed FUNC_{BEGIN,END} macros in comply with rc3
  * in setup.{py,cfg} updated version to 0.3.3-rc3

[2011/12/05] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c added __author__ from compile-time macro AUTHOR
  * in sandbox/module.c added signal_info field in the return of probe() and 
    pushed 'signal' field generation to the probe() method of the wapper class
    in sandbox/__init__.py
  * abandoned planned support of multi-process task

[2011/11/30] LIU Yu, <pineapple.liu@gmail.com>
  * moved version definition from __init__.py to compile-time macro VERSION
  * in setup.{py,cfg} updated version to 0.3.3-rc2

[2011/11/10] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c fixed 64-32 bit conversion error
  * in sandbox/module.c changed SandboxEvent.ext0 binding
  * in sandbox/__init__.py and setup.{py,cfg} updated version to 0.2.3-1

[2011/11/08] LIU Yu, <pineapple.liu@gmail.com>
  * in sandbox/module.c pushed most probe() fields to the probe() method of

sample2.py
#!/usr/bin/env python
################################################################################
# The Sandbox Libraries (Python) - Sample Script                               #
#                                                                              #
# Copyright (C) 2009-2013 LIU Yu, <pineapple.liu@gmail.com>                    #
# All rights reserved.                                                         #
#                                                                              #
# Redistribution and use in source and binary forms, with or without           #
# modification, are permitted provided that the following conditions are met:  #
#                                                                              #
# 1. Redistributions of source code must retain the above copyright notice,    #
#    this list of conditions and the following disclaimer.                     #
#                                                                              #
# 2. Redistributions in binary form must reproduce the above copyright notice, #
#    this list of conditions and the following disclaimer in the documentation #
#    and/or other materials provided with the distribution.                    #
#                                                                              #
# 3. Neither the name of the author(s) nor the names of its contributors may   #
#    be used to endorse or promote products derived from this software without #
#    specific prior written permission.                                        #
#                                                                              #
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"  #
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE    #
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE   #
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE     #
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR          #
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF         #
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS     #
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN      #
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)      #
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE   #
# POSSIBILITY OF SUCH DAMAGE.                                                  #
################################################################################

import os
import sys

try:
    # check platform type
    system, machine = os.uname()[0], os.uname()[4]
    if system not in ('Linux', ) or machine not in ('i686', 'x86_64', ):
        raise AssertionError("Unsupported platform type.\n")
    # check package availability / version
    import sandbox
    if not hasattr(sandbox, '__version__') or sandbox.__version__ < "0.3.4-3":
        raise AssertionError("Unsupported sandbox version.\n")
    from sandbox import *
except ImportError:
    sys.stderr.write("Required package(s) missing.\n")
    sys.exit(os.EX_UNAVAILABLE)
except AssertionError as e:
    sys.stderr.write(str(e))
    sys.exit(os.EX_UNAVAILABLE)


def main(args):
    # sandbox configuration
    cookbook = {
        'args': args[1:],               # targeted program
        'stdin': sys.stdin,             # input to targeted program
        'stdout': sys.stdout,           # output from targeted program
        'stderr': sys.stderr,           # error from targeted program
        'quota': dict(wallclock=30000,  # 30 sec
                      cpu=2000,         # 2 sec
                      memory=8388608,   # 8 MB
                      disk=1048576)}    # 1 MB
    # create a sandbox instance and execute till end
    msb = MiniSandbox(**cookbook)
    msb.run()
    # verbose statistics
    sys.stderr.write("result: %(result)s\ncpu: %(cpu)dms\nmem: %(mem)dkB\n" %
        msb.probe())
    return os.EX_OK


# mini sandbox with embedded policy
class MiniSandbox(SandboxPolicy, Sandbox):
    sc_table = None
    # white list of essential linux syscalls for statically-linked C programs
    sc_safe = dict(i686=set([0, 3, 4, 19, 45, 54, 90, 91, 122, 125, 140,
        163, 192, 197, 224, 243, 252, ]), x86_64=set([0, 1, 5, 8, 9, 10,
        11, 12, 16, 25, 63, 158, 219, 231, ]), )
    # result code translation table
    result_name = dict((getattr(Sandbox, 'S_RESULT_%s' % r), r) for r in
        ('PD', 'OK', 'RF', 'RT', 'TL', 'ML', 'OL', 'AT', 'IE', 'BP'))

    def __init__(self, *args, **kwds):
        # initialize table of system call rules
        self.sc_table = dict()
        if machine == 'x86_64':
            for (mode, abi) in ((0, 'x86_64'), (1, 'i686'), ):
                for scno in MiniSandbox.sc_safe[abi]:
                    self.sc_table[(scno, mode)] = self._CONT
        else:  # i686
            for scno in MiniSandbox.sc_safe[machine]:
                self.sc_table[scno] = self._CONT
        # initialize as a polymorphic sandbox-and-policy object
        SandboxPolicy.__init__(self)
        Sandbox.__init__(self, *args, **kwds)
        self.policy = self  # apply local policy rules

    def probe(self):
        # add custom entries into the probe dict
        d = Sandbox.probe(self, False)
        d['cpu'] = d['cpu_info'][0]
        d['mem'] = d['mem_info'][1]
        d['result'] = MiniSandbox.result_name.get(self.result, 'NA')
        return d

    def __call__(self, e, a):
        # handle SYSCALL/SYSRET events with local rules
        if e.type in (S_EVENT_SYSCALL, S_EVENT_SYSRET):
            scinfo = (e.data, e.ext0) if machine == 'x86_64' else e.data
            rule = self.sc_table.get(scinfo, self._KILL_RF)
            return rule(e, a)
        # bypass other events to base class
        return SandboxPolicy.__call__(self, e, a)

    def _CONT(self, e, a):  # continue
        a.type = S_ACTION_CONT
        return a

    def _KILL_RF(self, e, a):  # restricted func.
        a.type, a.data = S_ACTION_KILL, S_RESULT_RF
        return a


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.stderr.write("synopsis: python " + __file__ + " foo/bar.exe\n")
        sys.exit(os.EX_USAGE)
    sys.exit(main(sys.argv))
